Fix #12: allow custom window position config

idr4n opened the pull request:
Hi,

I just recently discovered this plugin which I really like, partially as I also have been working on a preview plugin but for netrw.

I normally use a nvim-tree floating window, therefore this issue was relevant for me as well.

This PR basically allows for the user to optionally define the position of the preview window (either row, col, or both) with a number or a function that optionally accepts `tree_win` and/or `size`.

I think this approach is flexible to solve issues with floating nvim-tree windows based on user preferences.

For example, this is my nvim-tree config which float the window to the right:

```lua
view.float = {
        enable = true,
        quit_on_focus_loss = false,
        open_win_config = function()
          local screen_w = vim.opt.columns:get()
          local screen_h = vim.opt.lines:get() - vim.opt.cmdheight:get()
          -- local window_w = screen_w * 0.3
          local window_w = math.min(math.floor(screen_w * 0.28), 30)
          local window_h = screen_h * 0.9
          local window_w_int = math.floor(window_w)
          local window_h_int = math.floor(window_h)

          -- adjust for the offset
          local col_right_aligned = screen_w - window_w_int - 3
          local row_offset = 1

          return {
            border = "rounded",
            relative = "editor",
            row = row_offset,
            col = col_right_aligned,
            width = window_w_int,
            height = window_h_int,
          }
        end,
}
```

with this PR, I can add this configuration in your plugin, and positions the preview window nicely (I just modify the column position as the row is nicely calculated by your plugin):

```lua
  dependencies = {
    {
      "idr4n/nvim-tree-preview.lua",
      opts = {
        win_position = {
          ---@param tree_win number? The tree window handle
          ---@param size {width: number, height: number} Desired window dimensions
          ---@return number Calculated column position
          col = function(tree_win, size)
            return -1 * (size.width + 3)
          end,
        },
      },
    },
  },
```

![CleanShot 2025-07-05 at 07 06 05@2x](https://github.com/user-attachments/assets/93c5410e-fb1f-49ba-89ba-d573c5eb2ec0)

Let me know what you think, I would be happy to make any changes that you consider relevant or needed, in addition to updating the README at the end.

Thanks in advance!

idr4n committed:
Fix #12: allow custom window position config

b0o commented on lua/nvim-tree-preview/preview.lua:
```diff
@@ -486,32 +486,56 @@ end
 
-  -- Calculate cursor position relative to visible window
-  local relative_cursor = cursor_row - topline
+  if calculate_row or calculate_col then
```
Rather than using the intermediate `calculate_row`/`calculate_col` variables, what if we just check if `row` or `col` is `nil` here?

idr4n replied:
Done! You are right! we can omit those intermediate variables.

b0o commented on lua/nvim-tree-preview/config.lua:
```diff
@@ -22,6 +22,7 @@ local M = {
     title_pos = 'top-center',
     title_format = ' %s ',
     follow_links = true,
+    win_position = {},
```
Let's add some validation for this down in the `setup` function.

idr4n replied:
Done!

b0o reviewed (commented):
Looks good. Thank you for this. I left a few comments, if we address those and update the README I'll merge it!

idr4n committed:
refactor: omit intermediate variables

idr4n committed:
refactor: add validation in config

idr4n committed:
doc: update README with win_position config

idr4n commented:
Thanks a lot for reviewing this! I have addressed your comments and updated the README. Let me know if you would like to change anything of the above modifications.

b0o commented:
Looks great, thank you for your contribution!

idr4n commented:
Thank you for merging this!
